pkg/restic/repository_manager.go

```diff
@@ -0,0 +1,574 @@
+	volumesToBackup := GetVolumesToBackup(pod)
+	if len(volumesToBackup) == 0 {
+		return nil
+	}
+
+	// ensure a repo exists for the pod's namespace
+	if err := rm.ensureRepo(pod.Namespace); err != nil {
+		return err
+	}
+
+	resultChan := make(chan *arkv1api.PodVolumeBackup)
+
+	informer := arkv1informers.NewFilteredPodVolumeBackupInformer(
+		rm.arkClient,
+		backup.Namespace,
+		0,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(opts *metav1.ListOptions) {
+			opts.LabelSelector = fmt.Sprintf("%s=%s", arkv1api.BackupNameLabel, backup.Name)
+		},
+	)
+	informer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(_, obj interface{}) {
+				pvb := obj.(*arkv1api.PodVolumeBackup)
+
+				if pvb.Status.Phase == arkv1api.PodVolumeBackupPhaseCompleted || pvb.Status.Phase == arkv1api.PodVolumeBackupPhaseFailed {
+					resultChan <- pvb
+				}
+			},
+		},
+	)
+	go informer.Run(ctx.Done())
+	if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+		return errors.New("timed out waiting for cache to sync")
+	}
+
+	for _, volumeName := range volumesToBackup {
+		rm.repoLocker.Lock(pod.Namespace, false)
+		defer rm.repoLocker.Unlock(pod.Namespace, false)
+
+		volumeBackup := &arkv1api.PodVolumeBackup{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    backup.Namespace,
+				GenerateName: backup.Name + "-",
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: arkv1api.SchemeGroupVersion.String(),
+						Kind:       "Backup",
+						Name:       backup.Name,
+						UID:        backup.UID,
+						Controller: boolptr.True(),
+					},
+				},
+				Labels: map[string]string{
+					arkv1api.BackupNameLabel: backup.Name,
+					arkv1api.BackupUIDLabel:  string(backup.UID),
+				},
+			},
+			Spec: arkv1api.PodVolumeBackupSpec{
+				Node: pod.Spec.NodeName,
+				Pod: corev1api.ObjectReference{
+					Kind:      "Pod",
+					Namespace: pod.Namespace,
+					Name:      pod.Name,
+					UID:       pod.UID,
+				},
+				Volume: volumeName,
+				Tags: map[string]string{
+					"backup":     backup.Name,
+					"backup-uid": string(backup.UID),
+					"pod":        pod.Name,
+					"pod-uid":    string(pod.UID),
+					"ns":         pod.Namespace,
+					"volume":     volumeName,
+				},
+				RepoPrefix: rm.repoPrefix,
+			},
+		}
+
+		// TODO should we return here, or continue with what we can?
+		if err := errorOnly(rm.arkClient.ArkV1().PodVolumeBackups(volumeBackup.Namespace).Create(volumeBackup)); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	var errs []error
+
+ForLoop:
+	for i := 0; i < len(volumesToBackup); i++ {
+		select {
+		case <-ctx.Done():
+			errs = append(errs, errors.New("timed out waiting for all PodVolumeBackups to complete"))
+			break ForLoop
+		case res := <-resultChan:
+			switch res.Status.Phase {
+			case arkv1api.PodVolumeBackupPhaseCompleted:
+				SetPodSnapshotAnnotation(pod, res.Spec.Volume, res.Status.SnapshotID)
```

Comment:
I still think I'd like to see this function return the pod/snapshot associations, and have another function modify that pod. It hopefully will make testing easier.